axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,212 @@ def strategy(self, opponent: Player) -> Action:
+        self.two_turns_after_good_defection_ratio= 0
+        self.one_turn_after_good_defection_ratio_count = 1
+        self.two_turns_after_good_defection_ratio_count = 1
+        self.score = None
+        self.probability = 1.1
+        self.current_score = 0
+        self.dict = {C: 0, D: 1}
+
+
+    def update_state(self, opponent):  
+        
+        """
+        Calculates the ratioFD values and P values, as well as sets the 
+        states of FD at the start of each turn
+        """
+        self.current_score = compute_final_score(zip(self.history, opponent.history))
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+            ((self.two_turns_after_good_defection_ratio 
+            * self.two_turns_after_good_defection_ratio_count) 
+            + (3 - (3 * self.dict[opponent.history[-1]])) 
+            + (2 * self.dict[self.history[-1]]) 
+            - ((self.dict[opponent.history[-1]] 
+            * self.dict[self.history[-1]]))) 
+            / (self.two_turns_after_good_defection_ratio_count + 1)
+            )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+            ((self.one_turn_after_good_defection_ratio 
+            * self.one_turn_after_good_defection_ratio_count)
+            + (3 - (3 * self.dict[opponent.history[-1]])) 
+            + (2 * self.dict[self.history[-1]]) 
+            - (self.dict[opponent.history[-1]] 
+            * self.dict[self.history[-1]])) 
+            / (self.one_turn_after_good_defection_ratio_count + 1)
+            )
+            self.one_turn_after_good_defection_ratio_count += 1
+        else:
+            if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+                self.probability = (
+                (.95 - (((self.one_turn_after_good_defection_ratio)
+                + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                + (1 / (((len(self.history))+1) ** 2))
+                - (self.dict[opponent.history[-1]] / 4)
+                )
+                self.probability = round(self.probability, 4)
+                self.score = "good"
+            elif (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+                self.probability = (
+                (.25 + (opponent.cooperations / ((len(self.history)) + 1)))
+                - (self.consecutive_defections * .25) 
+                + ((self.current_score[0] 
+                - self.current_score[1]) / 100) 
+                + (4 / ((len(self.history)) + 1)))
+                self.probability = round(self.probability, 4)
+                self.score = "average"
+
+    def strategy(self, opponent: Player) -> Action:
+
+        randomValue = random.random()
```

Comment:
`random_value` (PEP8 recommend snake case for variable names)

Reply:
Can we also move this random value sample to line 403 where I think it is needed. Otherwise we're sampling a number every time which is a minor extra computational cost.

Reply:
Good catch! I ended up just removing the entire `randomValue` and replaced it with `random.random()` as was only used in place.

Reply:
Great. That's a good way to do this.